docs/rfc/rfc-001-parallel-tx-processing.md

```diff
@@ -0,0 +1,121 @@
+
+Upon receiving a block proposal, the app will first build out the access DAG based on the rules and mapping aforementioned, and validate that the DAG is indeed a DAG, so that there can be no circular dependency that would make deadlock possible. Then it will fire off one goroutine per message with `resource_dependencies` and `completion_signals` set for each message based on the DAG. The app blocks on waiting for all goroutines before responding to Tendermint.
+#### Maintain Access Mapping
+The following structs will be defined to represent resource access:
+```go
+type AccessType uint32
+const (
+    AccessType_Unknown = 0
+    AccessType_Read = 1
+    AccessType_Write = 2
+)
+
+type ResourceType uint32
+const (
+    ResourceType_Unknown = 0
+    ResourceType_Any = 1 // The most coarse-grained resource type. Writes to it block everything.
+    ResourceType_KV = 2
+    ResourceType_Mem = 3
+    ResourceType_DexMem = 4
+    ...
+)
+
+type AccessOp struct {
+    MsgKey string
+    Type AccessType
+    Resource ResourceType
+    IdentifierTemplate string // can be generalized
+}
+
+func (a *AccessOp) GetResourceIdentifier(args []any) string {
+    return fmt.Sprintf(a.IdentifierTemplate, args...)
+}
+```
+Two new interface methods will be added to `sdk.Msg` which each message type would need to implement:
+```go
+GetAccessMappingKey() string
+GetAccessMappingValueArgs() [][]any
+```
+The new `accesscontrol` module needs to have mapping for every possible outcome of `GetAccessMappingKey()` in its store; otherwise the message should be rejected during CheckTx. Mapped value for each "mapping key" is an ordered list of `AccessOp`s. For example, assume the `i`-th `AccessOp` of a certain message type is a bank balance read operation:
+- the key may be something like `bank-balances`
+- the type is `read`
+- the resource type is `KV`
+- the id template can just be `"%s"`, whereas `GetAccessMappingValueArgs` of a particular instance of that message type will return a list `args` such that `args[i]` is `[<account address>]`.
+
+Note that if the exact value of a template argument cannot be known before actually executing the message, then it should not be a template and the mapping should be set with a coarser grained `AccessOp` (e.g. in this case, if account cannot be known in advance, then the `AccessOp` should cover the entire bank/balances prefix)
+
+Mappings for builtin modules are set in genesis and can only be updated via governance. If governance messages themselves have wrong mappings set, the chain may never be recovered, so for safety we will switch to the old sequential execution mode whenever there are governance messages in a block.
+
+Mappings for `wasm` contracts' `execute` endpoints can be kept coarse-grained for now, with the following `AccessOp`s definition (in order):
```

Comment:
nit: maybe add a section for how the DAG looks like (i.e. a table of the ordering of module/resources processed)